File: python/01-arrays-hashing/array_chunk.py
"""
Problem: Array Chunk
Pattern: Arrays & Hashing (list slicing)
Time: O(n) | Space: O(n)

Given a list and a chunk size, split it into sublists of that length
(the last chunk may be shorter).

Input:  chunk([1, 2, 3, 4, 5], 2)
Output: [[1, 2], [3, 4], [5]]
"""

def chunk(array, size):
    result = []
    index = 0
    while index < len(array):
        result.append(array[index:index + size])
        index += size
    return result


# Alternative: build with a list comprehension over range steps instead of
# a while-loop + slicing. Same O(n) time, more Pythonic one-liner.
def chunk_comprehension(array, size):
    return [array[i:i + size] for i in range(0, len(array), size)]


if __name__ == "__main__":
    print(chunk([1, 2, 3, 4, 5], 2))  # [[1, 2], [3, 4], [5]]

File: javascript/01-arrays-hashing/arrayChunk.js
/**
 * Problem: Array Chunk
 * Pattern: Arrays & Hashing (array slicing)
 * Time: O(n) | Space: O(n)
 *
 * Given an array and a chunk size, split it into subarrays of that length
 * (the last chunk may be shorter).
 *
 * Input:  chunk([1, 2, 3, 4, 5], 2)
 * Output: [[1, 2], [3, 4], [5]]
 */
function chunk(array, size) {
  const chunk = [];
  let index = 0;

  while (index < array.length) {
    chunk.push(array.slice(index, index + size));
    index += size;
  }
  
  return chunk;
}

// Alternative: build with reduce instead of a while-loop + slice.
// Same O(n) time, more functional style -- some interviewers like seeing both.
function chunkReduce(array, size) {
  return array.reduce((result, item, index) => {
    const chunkIndex = Math.floor(index / size);
    if (!result[chunkIndex]) result[chunkIndex] = [];
    result[chunkIndex].push(item);
    return result;
  }, []);
}

module.exports = chunk;
module.exports.alternative = chunkReduce;
